### PR TITLE
test backend wrapper

### PR DESCRIPTION
**Pull request by Roshrini**

### CODE REVIEW

File: onnx_mxnet/backend.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+# permissions and limitations under the License.
+
+# coding: utf-8
+from .import_onnx import GraphProto
+import mxnet as mx
+import numpy as np
+from onnx.backend.base import Backend, BackendRep
+from collections import namedtuple
+
+"""Using these functions for onnx test infrastructure."""
```
Comment (lupesko) [inline onnx_mxnet/backend.py]: Can you elaborate more in the comment about the onnx test infra, and what this class encapsulates?

File: onnx_mxnet/backend.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+class MXNetBackend(Backend):
+    @classmethod
+    def run_node(cls, node, inputs, device='CPU'):
+        """Running individual node inference on mxnet engine and
+        return the result to onnx test infrastructure."""
+        g = GraphProto()
```
Comment (lupesko) [inline onnx_mxnet/backend.py]: Can we use a more descriptive var name such as graph or graph_proto?

File: onnx_mxnet/backend.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                tuples = ((batch_size,), inputs[idx].shape)
+                new_shape = sum(tuples, ())
+                data_shapes.append((input_name, new_shape))
+            else:
+                data_shapes.append((input_name, inputs[idx].shape))
+
+        # create a module
+        mod = mx.mod.Module(symbol=sym, data_names=data_names, label_names=None)
+        mod.bind(for_training=False, data_shapes=data_shapes, label_shapes=None)
+        # initializing parameters for calculating result of each individual node
```
Comment (lupesko) [inline onnx_mxnet/backend.py]: nit: add newline to make the sections in the method more readable

File: onnx_mxnet/backend.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        mod.init_params()
+        from collections import namedtuple
```
Comment (lupesko) [inline onnx_mxnet/backend.py]: You already have this statement on the header

File: onnx_mxnet/backend.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import mxnet as mx
+import numpy as np
+from onnx.backend.base import Backend, BackendRep
+from collections import namedtuple
+
+"""Using these functions for onnx test infrastructure."""
+class MXNetBackend(Backend):
+    @classmethod
+    def run_node(cls, node, inputs, device='CPU'):
+        """Running individual node inference on mxnet engine and
```
Comment (lupesko) [inline onnx_mxnet/backend.py]: Can you please have the docstring describe method parameters as well?

File: onnx_mxnet/backend.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    def supports_device(cls, device):
+        """Supports only CPU for testing"""
+        return device == 'CPU'
+
+
+class MXNetBackendRep(BackendRep):
+    """Running model inference on mxnet engine and return the result
+     to onnx test infrastructure for comparison."""
+    def __init__(self, mxnet_symbol, params):
+        self.model = mxnet_symbol
```
Comment (lupesko) [inline onnx_mxnet/backend.py]: Maybe naming this instance variable self.mxnet_symbol or self.symbol is more appropriate?

File: onnx_mxnet/backend.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        self.params = params
+
+    def run(self, inputs, **kwargs):
+        """Run model inference and return the result"""
+        input_data = np.asarray(inputs[0], dtype=np.float32)
```
Comment (lupesko) [inline onnx_mxnet/backend.py]: How do we know dtype is always float?
Should we check the inputs arg?

File: onnx_mxnet/backend.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        # create module
+        mod = mx.mod.Module(symbol=self.model, data_names=['input_0'], context=mx.cpu(), label_names=None)
+        mod.bind(for_training=False, data_shapes=[('input_0', input_data.shape)], label_shapes=None)
+        mod.set_params(arg_params=self.params, aux_params=None)
+        # run inference
```
Comment (lupesko) [inline onnx_mxnet/backend.py]: nit: newline

File: onnx_mxnet/import_helper.py [hunk 9]
```diff
@@ -91,11 +91,17 @@ def _conv_transpose():
         disables=['output_shape'],
         custom_check=_dimension_constraint())
 
+def _change_eps_cudnn(attr):
+    """Limiting eps value to 1e-5 for cudnn batchnorm."""
+    if attr < 1e-5:
+        attr = 1e-4
+    return attr
+
 # converting attributes for BatchNorm operator
```
Comment (lupesko) [inline onnx_mxnet/import_helper.py]: Proper Docstring?

File: onnx_mxnet/import_helper.py [hunk 10]
```diff
@@ -106,13 +112,19 @@ def _activation(name):
             'alpha':'slope'},
         extras={'act_type': name})
 
-# [TODO] : Only 4d 5d tensor supported in mxnet.
+def _pad_sequence_fix(attr):
```
Comment (lupesko) [inline onnx_mxnet/import_helper.py]: Proper Docstring?

File: onnx_mxnet/import_onnx.py [hunk 11]
```diff
@@ -119,7 +119,11 @@ def from_onnx(self, graph):
             inputs = [self._nodes[self._renames.get(i, i)] for i in node.input]
             if op_name == 'Gemm':
```
Comment (lupesko) [inline onnx_mxnet/import_onnx.py]: Can you add a comment to explain the issue with Gemm?

File: onnx_mxnet/import_onnx.py [hunk 12]
```diff
@@ -134,6 +138,40 @@ def from_onnx(self, graph):
...
         return out, self._params
 
+    def _run_node(self, node, device='CPU'):
+        """Construct symbol from individual node.
+        Mainly using this function for unittests"""
+        op_name = node.op_type
+        attr = self._parse_attr(node.attribute)
+        new_op, new_attr = _convert_operator(op_name, attr)
+        sym_list = [mx.sym.Variable(node_name) for node_name in node.input]
+        # some workarounds for onnx problem
```
Comment (lupesko) [inline onnx_mxnet/import_onnx.py]: nit: newline

File: onnx_mxnet/import_onnx.py [hunk 13]
```diff
@@ -134,6 +138,40 @@ def from_onnx(self, graph):
...
+        new_attr = self._fix_bias(new_op, new_attr, len(sym_list))
+        new_attr = self._fix_channels(new_op, new_attr, list(node.input))
+        # calling again to get new symbols after some workarounds
```
Comment (lupesko) [inline onnx_mxnet/import_onnx.py]: nit: newline

File: onnx_mxnet/tests/onnx_backend_test.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+import onnx.backend.test
+from onnx_mxnet import backend as mxnet_backend
+
+# This is a pytest magic variable to load extra plugins
+pytest_plugins = 'onnx.backend.test.report',
+
+backend_test = onnx.backend.test.BackendTest(mxnet_backend, __name__)
+
+# will add model tests later
```
Comment (lupesko) [inline onnx_mxnet/tests/onnx_backend_test.py]: Please file an issue and link it here , this is a good way to ensure these kind of TODOs are tracked.

**Comment by Roshrini** [comment]
@nswamy

**Pull request merged**